src/components/__tests__/ProseMirror.draw.test.tsx

```diff
@@ -0,0 +1,263 @@
+    const events: string[] = [];
+    class PluginView {
+      update() {
+        events.push("update");
+      }
+      destroy() {
+        events.push("destroy");
+      }
+    }
+    const plugin = new Plugin({
+      view() {
+        events.push("create");
+        return new PluginView();
+      },
+    });
+    const { view, unmount } = tempEditor({ plugins: [plugin] });
+    act(() => {
+      view.dispatch(view.state.tr.insertText("u"));
+    });
+    unmount();
+    expect(events.join(" ")).toBe("create update destroy");
+  });
+
+  it("redraws changed node views", () => {
+    const { view, rerender } = tempEditor({ doc: doc(p("foo"), hr()) });
+    expect(view.dom.querySelector("hr")).toBeTruthy();
+    rerender({
+      nodeViews: {
+        horizontal_rule: forwardRef(function Var(
+          props: NodeViewComponentProps,
+          ref
+        ) {
+          return <var ref={ref}>{props.children}</var>;
+        }),
+      },
+    });
+    expect(!view.dom.querySelector("hr")).toBeTruthy();
+    expect(view.dom.querySelector("var")).toBeTruthy();
+  });
+
+  it("doesn't get confused by merged nodes", () => {
+    const { view } = tempEditor({
+      doc: doc(p(strong("one"), " two ", strong("three"))),
+    });
+    act(() => {
+      view.dispatch(view.state.tr.removeMark(1, 4, schema.marks.strong));
+    });
+    expect(view.dom.querySelectorAll("strong")).toHaveLength(1);
+  });
+
+  it("doesn't redraw too much when marks are present", () => {
+    const s = new Schema({
+      nodes: {
+        doc: { content: "paragraph+", marks: "m" },
+        text: { group: "inline" },
+        paragraph: schema.spec.nodes.get("paragraph")!,
+      },
+      marks: {
+        m: {
+          toDOM: () => ["div", { class: "m" }, 0],
+          parseDOM: [{ tag: "div.m" }],
+        },
+      },
+    });
+    const paragraphs = [];
+    for (let i = 1; i <= 10; i++)
+      paragraphs.push(
+        s.node("paragraph", null, [s.text("para " + i)], [s.mark("m")])
+      );
+    const { view } = tempEditor({
+      // @ts-expect-error this is fine
+      doc: s.node("doc", null, paragraphs),
+    });
+    const initialChildren = Array.from(view.dom.querySelectorAll("p"));
+    const newParagraphs: Node[] = [];
+    for (let i = -6; i < 0; i++)
+      newParagraphs.push(
+        s.node("paragraph", null, [s.text("para " + i)], [s.mark("m")])
+      );
+    act(() => {
+      view.dispatch(view.state.tr.replaceWith(0, 8, newParagraphs));
+    });
+    const currentChildren = Array.from(view.dom.querySelectorAll("p"));
+    let sameAtEnd = 0;
+    while (
+      sameAtEnd < currentChildren.length &&
+      sameAtEnd < initialChildren.length &&
+      currentChildren[currentChildren.length - sameAtEnd - 1] ==
+        initialChildren[initialChildren.length - sameAtEnd - 1]
+    )
+      sameAtEnd++;
+    // $$FORK: Our node stability isn't _quite_ as robust
+    // as prosemirror-view's. The node adjacent to the one
```

Comment:
do you have a sense for why this is? like: is this about marks specifically, or generally about updates to DOM nodes adjacent to a given replacement?

Reply:
Ugh, unfortunately I don't remember the exact details of why this unit test is like one less than what it "should" be, but I did spend about an hour determining that it was fine haha. It's related to the way we do DOM stability/React key assignment generally. I'll try to explain, but take a look at the react keys plugin, too. DOM element stability is challenging in ProseMirror, because node content is an array, and the only way to preserve DOM stability in an array of elements in React is with unique keys. Buuuuttt... ProseMirror nodes don't have unique keys, nor any real unique identifiers. They can only really be uniquely identifiable by their position, which obviously can change as a result of any step.

So we have a plugin that stores a map from node position to a unique key for that node, and we do our best to keep those keys stable across transactions. This is actually quite good at node stability; it only runs into a small amount of trouble when adding new nodes, as this will inherently produce two positions that map backwards to the same previous position (making it challenging to determine which should actually receive the key from that position).

Anyway, something in there results in a teeny tiny amount of additional repaints in specific edge cases, but as the rest of this test suite demonstrates, this implementation gets really really close to prosemirror-view in terms of preserving DOM nodes.

Reply:
Got it. So it's maybe not specific marks, but instead general to React elements at the end of a list, or something like that? As you said doesn't seem super critical, though it might be interesting at some point to poke into when/why it does/doesn't trigger, if only to understand if there's some way we could improve the React element keying to reduce the odds of repaint

Reply:
I feel confident that it's not specific to marks, and yeah more like something to do with something like what you said. And I think documenting that is absolutely a good call, as well as seeing if it can be improved even further!

Reply:
I also wouldn't be surprised if there's places where stability is better than in the standard ProseMirror editor view. Although, stability _in theory_ isn't necessarily the same as stability in common practice. There might be cases this handles better but that don't really occur in practice. I remember thinking the way ProseMirror handles this is pretty clever, something like matching prefix and suffixes of the sibling lists, if I remember correctly, no?